### PR TITLE
Add/mobile fullscreen image preview

### PR DESCRIPTION
**Pull request by cameronvoell**
## Description
This PR covers fullscreen image preview in the gutenberg mobile editor as specified in [gb-mobile issue 1286](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1286)

WP iOS PR for testing: https://github.com/wordpress-mobile/WordPress-iOS/pull/12834
fullscreenimage-sketch
Remaining tasks:
1. [x] Fade animation to fullscreen image preview from Image block editor
2. [x] Image Preview has correct aspect ratio in portrait and landscape
3. [x] Highlight border around image when image block is selected (same as video block)

Decided to commit behind a `__DEV__` flag until rotation issues are resolved. Tasks below will be in a future PR. 
_4.  Address dependency on Fix for Android modal rotation crash ([wp-mobile PR 10383](https://github.com/wordpress-mobile/WordPress-Android/pull/10383#issuecomment-521123642))
5.  (•••) Button over Image
6.  (•••) Button leads to BottomSheet with Edit and ViewFullscreen options_

## How has this been tested?

Tested as a submodule to gutenberg-mobile repo at latest develop ([cf27fba](https://github.com/wordpress-mobile/gutenberg-mobile/commit/cf27fba4ce3a8243450ee52b0909a3508a6efd4a) )
### Test steps:
1. Open article in the block editor.
2. Navigate to an image block with an image loaded, or add a new image block and add an image
3. Select the image block and notice the blue border shows around the image
4. Press the image while the image block is selected and notice the image will be viewed in fullscreen.
5. Press outside the boundaries of the image in fullscreen mode, or swipe the image down to exit the fullscreen image preview. 

*Note you can also test rotating the device while in fullscreen preview, but there is a known issue that will cause the image preview to close on rotate when testing in WordPress Android on a phone running API 28 - See WordPress Android issues [10227](https://github.com/wordpress-mobile/WordPress-Android/issues/10227) and [1333](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1333)

## Screenshots 
Fade in to fullscreen image preview:
fullscreen-preview-v0

## Types of changes

New feature - this code updates the image block of the mobile editor so that users can do a fullscreen preview of the image that they are adding. 

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [X] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/components/src/mobile/image-viewer/index.native.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+				animationOut={ "fadeOut" }
+				animationOutTiming={ 250 }
+				backdropTransitionInTiming={ 50 }
+				backdropTransitionOutTiming={ 50 }
+				backdropOpacity={ 1 }
+				supportedOrientations={['portrait', 'landscape']}
+				onBackdropPress={ this.props.onClose }
+				onBackButtonPress={ this.props.onClose }
+				onSwipe={ this.props.onClose }
+				swipeDirection="down"
```
Comment (Tug) [inline packages/components/src/mobile/image-viewer/index.native.js]: I think we could use `[ 'up', 'down' ]` or even `[ 'up', 'down', 'left', 'right' ]`. It feels a bit weird to me to only being able to dismiss it by swiping down.

Comment (Tug) [reply packages/components/src/mobile/image-viewer/index.native.js]: I guess we'll need to adjust the `panResponder` accordingly?

File: packages/components/src/mobile/image-viewer/index.native.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		this.fetchImageSize();
+	}
+
+	componentDidUpdate( prevProps ) {
+		if ( this.props.url !== prevProps.url ) {
+			this.fetchImageSize();
+		}
+	}
+
+    componentWillUnmount() {
```
Comment (Tug) [inline packages/components/src/mobile/image-viewer/index.native.js]: Minor, but we should be using tabs only. We have a pre-commit hook for linting and other checks. You'll probably need the dependencies installed in `gutenberg` for that to work (`npm install`)

Comment (cameronvoell) [reply packages/components/src/mobile/image-viewer/index.native.js]: Yea I noticed I have a couple styling hiccups in here. I'll be sure to fix those.

File: packages/components/src/mobile/image-viewer/index.native.js (hunk 1 above)
Comment (cameronvoell) [reply packages/components/src/mobile/image-viewer/index.native.js]: I'm sure this is possible, I'll take a look.

File: packages/components/src/mobile/image-viewer/index.native.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	onDimensionsChange(dims) {
+		this.fetchImageSize();
+	}
+
+	fetchImageSize() {
+		this.container = Dimensions.get("window")
+		Image.getSize( this.props.url, ( width, height ) => {
+			this.image = { width, height };
```
Comment (Tug) [inline packages/components/src/mobile/image-viewer/index.native.js]: It doesn't seem `this.image` is used for anything else that `this.calculateSize()` I think we can refactor a bit here.

File: packages/components/src/mobile/image-viewer/index.native.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+			isVisible,
+      		url = '',
+		} = this.props;
+    
+		const panResponder = PanResponder.create( {
+			onMoveShouldSetPanResponder: ( evt, gestureState ) => {
+				// Activates swipe down over child Touchables if the swipe is long enough.
+				// With this we can adjust sensibility on the swipe vs tap gestures.
+				if ( gestureState.dy > 3 ) {
+					gestureState.dy = 0;
```
Comment (Tug) [inline packages/components/src/mobile/image-viewer/index.native.js]: Curious, why do we need to mutate the `gestureState` here?

**Review by Tug — CHANGES_REQUESTED** [review]
Great work so far @cameronvoell !
I had a few comments on the code that I think we should address, otherwise this is very good for a v1 imo

File: packages/components/src/mobile/image-viewer/index.native.js (hunk 1 above)
Comment (Tug) [reply packages/components/src/mobile/image-viewer/index.native.js]: Good point! Yeah let's address that in a separate PR 👍

File: packages/block-library/src/image/edit.native.js [hunk 5]
```diff
@@ -256,6 +263,14 @@ class ImageEdit extends React.Component {
 			</BottomSheet>
 		);
 
+		const getImageViewer = () => (
+			<ImageViewer
+				isVisible={ this.state.showImageViewer }
+				onClose={ onImageViewerClose }
+				url ={ url }
```
Comment (Tug) [inline packages/block-library/src/image/edit.native.js]: linter is complaining about the space here

Comment (cameronvoell) [reply packages/block-library/src/image/edit.native.js]: Fixed in 97df6948f422ccacfd4997b0587907f41a89d0ae

**Comment by etoledom** [comment]
I just re-run the Lint job and it now succeeded.

It's interesting that the lint error is real, and it should be failing everywhere, but even locally it succeeds. Anyway don't mind this, I asked for fixing this lint issue here: https://github.com/WordPress/gutenberg/pull/17251/files#r320642481

**Comment by etoledom** [comment]
Hey @cameronvoell - I gave this PR a test run. It looks quite good.

The animation feels fluid and the image size seems perfect 🎉 
The border around the Image also pretty cool.

I agree that just swiping down to dismiss is good enough for this iteration. Also tapping outside to dismiss works fine. 👍 

There is one difference with how the Video block works:
When the caption is active, pressing on the video will first un-focus the caption field. Then another tap is needed to show the video full screen.

On the Image block, when the caption is focused, tapping on the Image will both unfocus the caption and show the image full screen at the same time. I don't think this is an issue and I'm inclined to think that it's not a blocker, specially having this behind the DEV flag. But it's worth considering.

There's also a little glitch on iOS when this happens (not very visible in the GIF). But again, I don't think it's a blocker.

image_full_screen

**Comment by cameronvoell** [comment]
> Hey @cameronvoell - I gave this PR a test run. It looks quite good.

Thanks!

> I agree that just swiping down to dismiss is good enough for this iteration. Also tapping outside to dismiss works fine. 👍

I've already tested adding swipe in multiple directions after the required update to react-native-modal and it it works well also. I think this would be a nice quick PR to try after this one. I'm also interested if the updated react-native-modal library could have any effect on our modal rotation issues, I have not tested that yet. 

> On the Image block, when the caption is focused, tapping on the Image will both unfocus the caption and show the image full screen at the same time. I don't think this is an issue and I'm inclined to think that it's not a blocker, specially having this behind the DEV flag. But it's worth considering.

Thomas's mockups had us adding a button over the image that opens up an additional menu that has the option for going to fullscreen in the next version of this change. That seems like a good time to fix this.

**Comment by etoledom** [comment]
> Thomas's mockups had us adding a button over the image that opens up an additional menu that has the option for going to fullscreen in the next version of this change.

Sounds great to me! 👍

File: packages/block-library/src/image/edit.native.js [hunk 6]
```diff
@@ -116,9 +118,11 @@ class ImageEdit extends React.Component {
 		} else if ( attributes.id && ! isURL( attributes.url ) ) {
 			requestImageFailedRetryDialog( attributes.id );
 		}
-
+		// eslint-disable-next-line no-undef
+		const enableFullscreen = __DEV__;
```
Comment (Tug) [inline packages/block-library/src/image/edit.native.js]: We should cast it to a boolean here to avoid the `enableFullscreen && true` trick

**Review by Tug — APPROVED** [review]
LGTM! Let's 🚢 it.

**Comment by cameronvoell** [comment]
Closed in favor of native solution. More info: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1286#issuecomment-557195843

**Pull request closed without merging**